### ISSUE TITLE
Allow users to get info for repositories they aren't watching but also aren't ignoring

### ISSUE THREAD

**Issue by akalman** [ptsochantaris/trailer#299]
GitHub's setting for level of watching on a repository has three levels
- Watching (receive email notifications for PRs and issues etc on repository)
- Not Watching (receive email notifications on threads you are participating in or are mentioned in)
- Ignored (receive no emails)

currently in order to see a PR/Issue in trailer I need to be "Watching" the repository.  Each repository I watch I can filter by involvement to one of four levels
- Hide
- Mine
- Participated
- All

Users should be able to see PR's and Issues for repositories they are "Not Watching" and filter that repository to one of the first three options (excluding the "All" option).

**Comment by ptsochantaris** [comment]
Hi @akalman and sorry for taking so long to answer here. Trailer used to work this way very early on, where you could add and remove repositories ad-hoc instead of using the server watchlist. To make things simpler and more straightforward, it was simplified to simply following the watchlist 1:1.

However, as times goes on, and especially as GitHub adds more notification-oriented/side-effect functionality to the watchlist itself, I'm starting to think that what you say makes more sense, i.e. that while the watchlist is a great starting point as the basis of Trailer's own "watchlist", Trailer should perhaps not be "locked" to it.

What I'm thinking is to modify the "watchlist" section as follows:
- Initially when Trailer is set up, sync the watchlist just like now (i.e. 1:1)
- Option to "Automatically add repos which I add to my GitHub watchlist" (default yes)
- Option to "Automatically remove repos I remove from my GitHub watchlist" (default yes)
- A new dialog to add a repo outside of the watchlist (which is never touched by the above options)

So, the defaults would mirror Trailer's existing behaviour, keep the ease of setup, but you could also use the watchlist as a "bootstrap" instead of a continuous "synced" list, and tweak it as you like going forward.

What do you think of the above? Let me know your thoughts. Suggestions, as always, are welcome.

**Comment by akalman** [comment]
The behavior I really want is just the third bullet. being able to add repositories from outside the watchlist.  The ability to modify behavior when the watchlist is mutated would be great but I don't think it's really currently needed. I can see complications with a system like that as well.

Obviously if you want to go ahead and implement those features I won't complain but waiting until people have some time with repos outside of the watchlist before we really know what behavior we'd like when the watchlist is modified.

I also feel like there is benefit from divorcing the trailer repositories from github's watch setting even further. setting up several reasons why a repository would show up in trailer, and the watchlist being only one of those. for example I use trailer as my sole notification system for github activity. all my github email is filtered out of my inbox (not deleted just moved) so what i really need is all of my organizations repositories to show up in trailer. something like "show my watchlist and this organization".

But the bottom line is right now i just need some way to watch repositories outside of my watchlist. it doesn't have to be perfect because i don't know what perfect is. being able to add repositories ad hoc would certainly meet that criteria

**Comment by ptsochantaris** [comment]
Thanks for the feedback @akalman - and yes, I agree with most of what you say. However, there are some reasons why I'm suggesting this "package" of changes, and I feel describing them in more detail may address some of your concerns.

- First is because the proposed option mentally divorces the 1:1 mapping of watchlist-Trailer repos, then the perceptual model changes for the user. The "Watchlist" screen is no longer literally your watchlist (in GitHub terms, I mean) - it becomes "a list of repos that you are watching in Trailer". The fact that this list will (in many cases) identically match your GitHub watchlist is now a side-effect, and not an exact description of what it represents anymore. The proposed options for automatically following the repo changes will ensure that the behaviour of this more generic view will remain unaltered, and that for the "specialised" use case (i.e. just mirroring the watchlist) no new habits or conceptual models need to change in the user's mental model.
- To put it another way, the thing we don't want, either in code or the user's mind, is to create a "caste" system of repos, where one caste is "managed by the watchlist" and another one is "manually managed". Just having an option to add special unmanaged repos into the list would create such an implicit "class divide". This divide would lead to UX issues. For instance, why can I remove this repo and not that other, etc? (We could have two lists, but *no*) Instead, with the new options, they will now all be equals, i.e. manually manageable, with the same options and UI behaviour -- EXCEPT that we introduce two new (default on) options to convert the previous *logic* into an *automation*. This way, the repositories view will be a flat repo list that can be optionally (and by default) auto-trimmed and added-to by using the GitHub watchlist. (For extra geekness, I think my argument for dealing with this is the same argument that others make for protocol-orientation vs inheritance, but *oh boy* am I off track now and this is getting too long :-P)

(Edit: Obviously, internally the manually-added repos would need a special flag for identification and exclusion from the watchlist auto-removal logic, and new watchlist repos that are the same as a manually-added repo would need some merging logic, but that's not something the *user* would need to think or worry about)

- The second reason has to do with UX. The interface for adding/removing ad-hoc repos from Trailer will need it's own space, most probably be part of an "advanced" popup-from-a-button in the repositories list that will most probably replace (and swallow) the "refresh watchlist now" button inside of it. Since we're going to be "exploding" the UI out in this direction, then we can certainly afford to add the two new options as well, and hence round out that view as a "repository list control panel", so to speak.

- Finally, these two options to "auto-add" and "auto-remove" are already implemented in the code. Or rather, I should say, they are implemented as logic that's already active, in order to sync the watchlist, right? Hence, the work to add a flag to basically "turn them off" would be near trivial. Basically, the app wouldn't bother running through that code-path if those checkboxes are off.

I hope this makes some sense and isn't too convoluted as a description :) Just wanted to give you a better idea behind my reasoning of why I'd rather go with a superset of features rather than the one single change you were taking about. To sum up: It's far easier for the user because it allows them to stay with a single-class/flat model of the repositories list, we will have the space for it in the UI, it will pad-out what may be an empty "advanced" dialog, *and* the extra options won't cost anything during implementation.

However, you did mention complications that you can see arising from such options, which is very interesting - care to elaborate?

**Comment by akalman** [comment]
Yeah that makes sense.  It sounds like the set of features isn't as big of a cost over just the add as I thought.  And like I said before, the feature set you listed definitely meets my needs, it was just maybe more than necessary.  You've brought up some good points why it isn't more than necessary.

The complications that immediately came to mind revolved around api usage (querying to get watchlist more frequently) and state management (needing to keep the previous watchlist saved to detect changes).  It sounds like the former is already happening and the latter is broadcasted, so they seem like non-issues.

Bottom line I do think the full feature set is an improvement over just the "add" functionality.  If it isn't super costly to do so then I don't see a problem with it.

**Comment by pklingem** [comment]
do I understand this discussion correctly that this enhancement would allow me to be notified when I am mentioned in a PR/issue on a repo that I am not watching?  If so I am excited about this possibility, it is the only remaining item for me to completely ignore github emails and rely on Trailer.  Trailer is truly a model of success for open source software, thank you so much for your contributions.

**Comment by ptsochantaris** [comment]
Hi @pklingem, yes, or at least the idea is to be able to supplement your watchlist in Trailer rather than have to specifically watch the repo in Github. Of course, this would mean that you have permission to view the Repo in the first place, but I think that goes without saying anyway.

This feature is already overdue if I were to go by my own estimate of when I would have started work on it, but unfortunately there are various reasons why I haven't gotten round to it yet. Rest assured, it is next on the TODO list :) and things will speed up again soon.

Finally, thank you so much for your comments on Trailer, I'm really glad to hear that it's making people's lives easier!

**Comment by ptsochantaris** [comment]
Hi! So, this is a build containing this feature. The "watchlist" section is now called "repositories" and contains an "advanced" button on the bottom-right. From there you can access the options mentioned above, as well as add or remove custom repositories which will not be affected by the watchlist, as long as you have access to them. Have a go and let me know your thoughts.

Feedback is very welcome, especially the removal panel which is currently a dropdown - it may be better to have a table of the custom-added repos from which the user can then select and remove entries. Depends on how many repos you want to follow outside our watchlist, I guess.

Please note that while this build seems rather stable, as always there may be dragons, so be sure to export your Trailer settings before installing it, in case you have to wipe it clean and downgrade to the release version because of some issue or crasher.

That said, the feature is pretty simple logic-wise, so I don't expect too many bumps.

https://www.dropbox.com/s/vxk8jzsb38satlq/Trailer-1.5.4-rc.zip?dl=0

**Comment by ptsochantaris** [comment]
Version 1.5.4 which is currently up should contain this feature now. I'm closing this issue but please feel free to open new ones with more feature requests, tweaks, etc.
